signer/apk/jar.go

```diff
@@ -248,3 +249,50 @@ func isSignatureFile(name string) bool {
 	}
 	return false
 }
+
+// isJARAligned checks if an APK is aligned to 4 bytes / 32-bits (to
+// which "allows those portions to be accessed directly with mmap()
+// even if they contain binary data with alignment restrictions.")
+// returns an err when an uncompressed file isn't aligned or another
+// error occurs (error reading ZIP bytes or fetching a file data
+// offset) should be equivalent to zipalign -c -v 4 from:
+// https://android.googlesource.com/platform/build.git/+/android-4.2.2_r1/tools/zipalign/README.txt
+func isJARAligned(input []byte) error {
+	const alignment = 4
+	r, err := zip.NewReader(bytes.NewReader(input), int64(len(input)))
+	if err != nil {
+		return err
+	}
+
+	for _, f := range r.File {
+		offset, err := f.DataOffset()
+		if err != nil {
+			return err
+		}
+		isCompressed := f.Method != zip.Store
+
+		if isCompressed {
+			log.Printf("%10d %s (OK - compressed)\n", offset, f.Name)
+			continue
+		}
+		if offset % alignment == 0 {
+			log.Printf("%10d %s (OK)\n", offset, f.Name)
+		} else {
+			log.Printf("%10d %s (BAD - %d)\n", offset, f.Name, offset % alignment)
+			return errors.Errorf("apk: unaligned file at %d %s (BAD - %d)", offset, f.Name, offset % alignment)
+		}
```

Comment:
Will those log statement will corrupt mozlogs? At any rate, that seems like a lot of noise we don't want in production.

Reply:
See comments on https://github.com/mozilla-services/autograph/pull/153#discussion_r223429211